graphql-lint/src/lib.rs

```diff
@@ -0,0 +1,654 @@
+                    self.diagnostics.push((
+                        format!("field '{field_name}' on type Subscription has a forbidden prefix: 'subscription'"),
+                        Severity::Warning,
+                    ));
+                }
+                if field_name.ends_with("Subscription") {
+                    self.diagnostics.push((
+                        format!("field '{field_name}' on type Subscription has a forbidden suffix: 'Subscription'"),
+                        Severity::Warning,
+                    ));
+                }
+            }
+            _ => {}
+        }
+    }
+
+    pub fn visit_directive(&mut self, directive: &DirectiveDefinition<'a>) {
+        if let CaseMatch::Incorrect { current, fix } = Self::case_check(directive.name(), Case::Camel) {
+            self.diagnostics.push((
+                format!("directive '{current}' should be renamed to '{fix}'"),
+                Severity::Warning,
+            ));
+        }
+    }
+
+    pub fn visit_directive_usage(&mut self, parent: &TypeDefinition<'a>, directive: &Directive<'a>) {
+        if directive.name() == "deprecated" && !directive.arguments().any(|argument| argument.name() == "reason") {
+            self.diagnostics.push((
+                format!(
+                    "usage of directive 'deprecated' on {} '{}' does not populate the 'reason' argument",
+                    Self::type_definition_display(parent),
+                    parent.name()
+                ),
+                Severity::Warning,
+            ));
+        }
+    }
+
+    pub fn visit_directive_usage_field(
+        &mut self,
+        parent_type: &TypeDefinition<'a>,
+        parent_field: &FieldDefinition<'a>,
+        directive: &Directive<'a>,
+    ) {
+        if directive.name() == "deprecated" && !directive.arguments().any(|argument| argument.name() == "reason") {
+            self.diagnostics.push((
+                format!(
+                    "usage of directive 'deprecated' on field '{}' on {} '{}' does not populate the 'reason' argument",
+                    parent_field.name(),
+                    Self::type_definition_display(parent_type),
+                    parent_type.name()
+                ),
+                Severity::Warning,
+            ));
+        }
+    }
+
+    pub fn visit_directive_usage_input_value(
+        &mut self,
+        parent_input: &InputObjectDefinition<'a>,
+        parent_input_value: &InputValueDefinition<'a>,
+        directive: &Directive<'a>,
+    ) {
+        if directive.name() == "deprecated" && !directive.arguments().any(|argument| argument.name() == "reason") {
+            self.diagnostics.push((
+                format!(
+                    "usage of directive 'deprecated' on input_value '{}' on input '{}' does not populate the 'reason' argument",
+                    parent_input_value.name(),
+                    parent_input.name()
+                ),
```

Comment:
Might be a nicer error if we use `Type.field` syntax here?
```suggestion
                format!(
                    "usage of directive 'deprecated' on '{}.{}' does not populate the 'reason' argument",
                    parent_input.name()
                    parent_input_value.name(),
                ),
```

Reply:
I considered that but it can be less descriptive if there's similarly named items of different types (you lose the type name)

Reply:
But we can consider doing it anyhow

Reply:
How do you lose the type name in `Type.field`?  It's before the dot?

Reply:
I mean you lose the description that it's an `input`, `type`, `interface` etc.